### ISSUE TITLE
[tests/docs] Update tests, example configs, and design documentation

### ISSUE THREAD

**Issue by ozw4** [ozw4/seisai#34]
Background

fbpick-coarse は、legacy local-crop / tiled inference 前提の設計から、global-anchor resize mode へ移行する。

新しい coarse pipeline の固定契約は以下。

input mode: global_anchor_resize

input shape:
  (3, 256, 2048)

target shape:
  (1, 256, 2048)

channels:
  0: waveform
  1: offset_ch
  2: time_ch

主な設計変更:

train / validation:
  full gather
  → gap-aware trace anchor selection
  → time resampling
  → fixed 3×256×2048 input

raw inference:
  full gather
  → deterministic center anchors
  → fixed 3×256×2048 input
  → anchor prediction
  → original trace / sample coordinates
  → .coarse.npz

この issue では、実装完了後に必要な tests / example configs / design docs を整理・更新する。

Goal

global-anchor coarse pipeline の仕様が tests と docs に明確に反映され、今後の変更で legacy tiled coarse 前提へ戻ったり、座標系・shape contract が壊れたりしないようにする。

更新対象:

- unit tests
- smoke tests
- example YAML
- README / docs
- migration notes
- design documentation
Scope
1. Update fbpick-coarse test expectations to global-anchor fixed shape

既存 test に 128×6016 / tiled / local crop 前提が残っている場合、global-anchor contract に更新する。

新しい期待 shape:

input:  (3, 256, 2048)
target: (1, 256, 2048)
logits: (B, 1, 256, 2048)

対象候補:

packages/seisai-engine/tests/
packages/seisai-dataset/tests/
tests/

確認したい contract:

- train dataset output shape
- validation dataset output shape
- raw inference dataset output shape
- model forward input shape
- logits shape validation
- .coarse.npz output shape
2. Add or update tests for global-anchor train / validation datasets

train dataset:

- draw_full_gather を使う
- anchor_mode="random"
- input shape is (3, 256, 2048)
- target shape is (1, 256, 2048)
- pad rows are masked
- anchor metadata exists at the expected location

validation dataset:

- anchor_mode="center"
- deterministic
- input shape is (3, 256, 2048)
- target shape is (1, 256, 2048)
- infer_endian is used for validation / infer files

必要であれば、mock / fixture を使って draw_full_gather が呼ばれることを検証する。

3. Add or update tests for raw global-anchor inference

raw inference test では、legacy tiled inference path が使われないことを確認する。

確認項目:

- raw inference dataset returns input shape (3, 256, 2048)
- raw inference dataset has no target
- raw inference uses center anchors
- raw inference is deterministic
- model forward receives (B, 3, 256, 2048)
- tiled-W inference utilities are not called

legacy tiled utility の例:

InferenceGatherWindowsDataset
collate_pad_w_right
TiledWConfig
iter_infer_loader_tiled_w
tile_w
overlap_w
tiles_per_batch

可能であれば monkeypatch で、これらが呼ばれたら fail する test を追加する。

4. Add tests for coordinate projection and restoration

global-anchor coarse では、座標変換が重要なので明示的に test する。

対象:

raw sample index → coarse sample index
coarse sample index → raw sample index
anchor prediction → full trace prediction
coarse_pick_i → coarse_pick_t_sec

期待:

raw_time_len = 6016
coarse_time_len = 2048

raw index 0    → coarse index 0
raw index 6015 → coarse index 2047

coarse index 0    → raw index 0
coarse index 2047 → raw index 6015

.coarse.npz 保存前後で以下も確認する。

coarse_pick_i.shape == (n_traces,)
coarse_pick_t_sec.shape == (n_traces,)
coarse_pmax.shape == (n_traces,)

coarse_pick_t_sec == coarse_pick_i * dt_sec
5. Add tests for gap-aware behavior

offset gap をまたいで情報が混ざらないことを test する。

確認項目:

- offset gap creates multiple segments
- anchor bin does not cross a gap
- train random anchors stay inside each segment
- validation / infer center anchors stay inside each segment
- prediction restoration interpolates only within each segment
- no interpolation occurs across offset gaps

人工 offset 例:

offsets = [0, 10, 20, 30, 1000, 1010, 1020]

期待:

segments:
  [0, 4)
  [4, 7)
6. Add tests for partial infer pair validation

config validation の regression test を追加する。

ケース:

A. infer_segy_files only
   → ValueError

B. infer_fb_files only
   → ValueError

C. both omitted
   → reuse training segy/fb pairs

D. both provided
   → use explicit infer segy/fb pairs

期待 error message:

paths.infer_segy_files and paths.infer_fb_files must be provided together
7. Update checkpoint metadata tests

global-anchor checkpoint metadata が保存・検証されることを確認する。

期待 metadata:

coarse_input_mode == "global_anchor_resize"
coarse_trace_len == 256
coarse_time_len == 2048
coarse_in_chans == 3
coarse_input_channels == ["waveform", "offset_ch", "time_ch"]

infer 側では、以下が失敗することを test する。

- metadata missing checkpoint
- legacy tiled checkpoint
- coarse_trace_len != 256
- coarse_time_len != 2048
- coarse_in_chans != 3
8. Update example train config

対象候補:

examples/config_train_fbpick_coarse.yaml

global-anchor coarse 前提に更新する。

例:

coarse:
  input_mode: global_anchor_resize

transform:
  trace_len: 256
  time_len: 2048
  standardize_eps: 1.0e-8

trace_anchor:
  gap_ratio: 5.0
  min_gap_m: null
  train_mode: random
  infer_mode: center

legacy local crop 前提の項目が残っている場合は削除またはコメントで非推奨化する。

例:

subset_traces
time crop specific settings
legacy coarse crop settings

ただし他 pipeline がまだ使う項目は削除しない。

9. Update example infer config

対象候補:

examples/config_infer_fbpick_coarse.yaml

global-anchor inference 前提に更新する。

例:

coarse:
  input_mode: global_anchor_resize

transform:
  trace_len: 256
  time_len: 2048
  standardize_eps: 1.0e-8

trace_anchor:
  gap_ratio: 5.0
  min_gap_m: null
  infer_mode: center

infer:
  ckpt_path: REPLACE_ME_coarse_best.pt
  device: auto
  batch_size: 1
  num_workers: 0
  amp: false
  use_tqdm: true

legacy tiled inference 専用の設定を削除または明確に deprecated 扱いにする。

tile_w
overlap_w
tiles_per_batch
overlap_h
10. Update README / docs for coarse global-anchor design

対象候補:

README.md
docs/
docs/fbpick*.md
docs/design*.md

記載する内容:

- fbpick-coarse の目的
- global-anchor resize mode の概要
- fixed input contract
- trace anchor selection の意味
- offset gap handling
- time resampling
- time channel semantics
- checkpoint metadata
- .coarse.npz output contract
- downstream physics / fine との関係

特に以下は明記する。

coarse は sample-level final pick ではなく、
fine / physics の探索中心を与える大局的位置推定 stage である。
11. Document time channel semantics

time channel は誤解が起きやすいため、明確に記載する。

正しい仕様:

time_view_sec は raw seconds grid
build_time_channel は正規化しない
time normalization は既存 BuildPlan の NormalizeTimeByConst が行う

既存 semantics:

time_ch = clip(time_view_sec / time_ref_sec, 0.0, 1.5)

禁止または非推奨:

time_view_sec[-1] で normalize して各 gather の最終 time を 1.0 にすること
12. Document .coarse.npz contract

global-anchor inference でも .coarse.npz は元 SEG-Y 座標で保存することを明記する。

主要 field:

coarse_pick_i:
  original SEG-Y sample index
  shape (n_traces,)

coarse_pick_t_sec:
  coarse_pick_i * dt_sec
  shape (n_traces,)

coarse_pmax:
  trace-wise confidence
  shape (n_traces,)

補足:

- internal model prediction is on 256×2048 coarse grid
- saved output is restored to original trace/sample coordinates
- physics / fine stages should not need schema changes
13. Add migration notes from legacy tiled coarse

legacy 設計との差分を docs にまとめる。

例:

Before:
  local crop / tiled inference
  H-window + W-tile
  arbitrary input size via tiling

After:
  full gather context
  gap-aware anchor selection
  fixed 3×256×2048 input
  deterministic center anchors for inference
  segment-wise restoration to original trace coordinates

注意点:

- old tiled coarse checkpoints are not compatible
- checkpoint metadata validation rejects legacy checkpoints
- raw inference no longer uses tile_w / overlap_w
- infer batch_size may initially be limited to 1 if metadata batching is not implemented
14. Update docs for evaluation expectations

coarse の評価は final sample accuracy だけでなく、後段 window に入るかを重視する。

記載例:

Recommended coarse metrics:
  - P50 / P90 / P95 absolute error
  - coverage within fine-window half width
  - gap-neighborhood error
  - trace-wise failure rate

特に重要:

|coarse_pick_i - fb_i| <= fine_window_half

この coverage を coarse stage の主要評価指標として推奨する。

Out of scope

この issue では以下は行わない。

- model architecture の変更
- train / validation dataset 実装の変更
- raw inference 実装の変更
- coordinate restoration 実装の変更
- loss の変更
- physics / fine stage の仕様変更
- viewer の大規模変更
- performance tuning

この issue は tests / configs / docs の整備に限定する。

Acceptance criteria
fbpick-coarse の tests が global-anchor fixed shape contract に更新されている。
train dataset の shape / random anchor / pad row behavior が test されている。
validation dataset の shape / deterministic center anchor behavior が test されている。
raw inference dataset / loop の global-anchor behavior が test されている。
tiled inference utilities が global-anchor path で使われないことが test されている。
coordinate projection / restoration の unit tests がある。
gap-aware anchor / restoration behavior の tests がある。
partial infer pair validation の regression tests がある。
checkpoint metadata validation tests がある。
example train config が global-anchor coarse 前提に更新されている。
example infer config が global-anchor coarse 前提に更新されている。
docs に global-anchor coarse design が記載されている。
docs に time channel semantics が明記されている。
docs に .coarse.npz contract が明記されている。
docs に legacy tiled coarse からの migration notes がある。
既存 tests が通る。
docs / examples に legacy tiled coarse 前提の誤解を招く記述が残っていない。
Suggested tests
1. Fixed train dataset shape
sample["input"].shape == (3, 256, 2048)
sample["target"].shape == (1, 256, 2048)
2. Fixed validation dataset shape
sample["input"].shape == (3, 256, 2048)
sample["target"].shape == (1, 256, 2048)
3. Validation deterministic anchors

同じ validation sample を複数回読む。

期待:

anchor_source_pos is identical
input is identical or allclose
target is identical or allclose
4. Train random anchors

同じ train sample を異なる seed / epoch context で読む。

期待:

anchor_source_pos may differ
shape contract remains fixed
5. Raw inference fixed shape
sample["input"].shape == (3, 256, 2048)
"target" not in sample
6. No tiled inference call

monkeypatch で tiled inference helper を呼ばれたら fail にする。

期待:

global-anchor infer path does not call tiled inference helper
7. Projection endpoint test
raw_time_len = 6016
coarse_time_len = 2048

raw 0    → coarse 0
raw 6015 → coarse 2047

coarse 0    → raw 0
coarse 2047 → raw 6015
8. Segment restoration does not cross gap

2 segment の人工 metadata を使う。

期待:

segment 0 traces use only segment 0 anchors
segment 1 traces use only segment 1 anchors
9. .coarse.npz smoke test

dummy model / fixture で inference を実行し、保存ファイルを読む。

期待:

coarse_pick_i.shape == (n_traces,)
coarse_pick_t_sec.shape == (n_traces,)
coarse_pmax.shape == (n_traces,)
coarse_pick_t_sec == coarse_pick_i * dt_sec
10. Checkpoint metadata mismatch test

legacy checkpoint または metadata 欠落 checkpoint を使う。

期待:

ValueError

message includes:

expected coarse_input_mode='global_anchor_resize'
Documentation outline

新規 docs を作る場合の候補:

docs/fbpick_coarse_global_anchor.md

構成案:

# FBPick Coarse Global-Anchor Resize Design

## Purpose
coarse stage の役割

## Input Contract
3×256×2048
waveform / offset_ch / time_ch

## Train / Validation Flow
full gather
gap-aware random / center anchors
time resampling
target construction

## Inference Flow
center anchors
model forward
anchor prediction
coordinate restoration
.coarse.npz

## Offset Gap Handling
segment split
no cross-gap pooling
no cross-gap interpolation

## Time Axis and Time Channel
raw seconds grid
NormalizeTimeByConst
no time_view_sec[-1] normalization

## Checkpoint Compatibility
global_anchor_resize metadata
legacy checkpoint rejection

## Migration from Legacy Tiled Coarse
removed / deprecated config fields
behavior changes

## Evaluation Recommendations
coverage within fine-window tolerance
gap-neighborhood errors
